docs/dev/adding_instruments.rst

```diff
@@ -732,6 +644,102 @@ In the above example, :code:`MultimeterA` and :code:`MultimeterB` use a differen
+
+Similarly, :meth:`Instrument.binary_values <pymeasure.instruments.Instrument.binary_values>` sends a command via :meth:`write`, waits with :meth:`wait_till_read`, but reads the response via :meth:`Adapter.read_binary_values <pymeasure.adapters.Adapter.read_binary_values>`.
+
+
+Adding a device address
+***********************
+
+Let's look at a simple example for a device, which requires its address as the first three characters and returns the same style. This is straightforward, as :meth:`write` just prepends the device address to the command, and :meth:`read` has to strip it again doing some error checking. Similarly, a checksum could be added.
+
+.. code-block:: python
+
+    class Extreme5002(Instrument):
+        """The great Extreme5002 instrument.
+        :param address: The device address for the communication.
+        """
+        def __init__(self, adapter, address=0):
+            super().__init__(adapter, "Extreme5002")
+            self.address = f"{address:03}"
+
+        def write(self, command):
+            """Add the device address in front of every command and send it to the device."""
+            super().write(self.address + command)
+
+        def read(self):
+            """Read from the device and assert, that the response starts with the device address."""
+            got = super().read()
+            if got.startswith(self.address):
+                return got[3:]
+            else:
+                raise ConnectionError(f"Message {got} for wrong address read.")
+    
+        voltage = Instrument.measurement(
+            ":VOLT:?", "The measured voltage in Volts.")
+
+If the device is initialized with :code:`address=12`, a request for the voltage would send :code:`"012:VOLT:?"` to the device and expect a response beginning with :code:`"012"`.
+
+
+Bytes communication
+*******************
+
+Some devices do not expect ASCII strings but raw bytes. In those cases, you can call the :meth:`write_bytes` and :meth:`read_bytes` in your :meth:`write` and :meth:`read` methods. The following example shows an instrument, which has registers to be written and read via bytes sent.
+
+.. code-block:: python
+
+    class Extreme5003(Instrument):
+    """The Extreme5003 instrument with bytes communication."""
+        def __init__(self, adapter):
+            super().__init__(adapter, "Extreme5003")
+    
+        def write(self, command):
+            """Write to the device according to the comma separated command.
+    
+            :param command: R or W for read or write, hexadecimal address, and data.
+            """
+            function, address, data = command.split(",")
+            b = [0x03] if function == "R" else [0x10]
+            b.append(int(address, 16))  # register address
+            b.extend(int(data).to_bytes(length=8, byteOrder="big", signed=True)
```

Comment:
Is that missing a close parens?
Should/can we doctest the new examples? Look at how we use (hidden) FakeInstrument above to achieve this, maybe this can still be leveraged here.

Reply:
I'll try to make it work, as I have not yet worked with doctests. Maybe `expected_protocol` will do.

Reply:
basically, using doctests you can, with a little "hidden" patching, test that the code examples/snippets you show in the docs actually produces the output shown, so you notice when the docs go out of sync with your code. See https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html

Reply:
I tried it, but generating the code with sphinx did not fail me despite obvious errors. As a stop gap measure, I tested the examples in an external file and added them afterwards again.

Reply:
Resolved.